---
layout: reveal_markdown
title: "Genomic data standards"
tags: slides 
date: 2021-12-09
---



# {{ page.title }}
---

## What is a standard?

> A norm or specification applied to a repeated process. It includes a formal definition of terms and a description of formats and expected behavior of a system

<p class="fragment">Standards &rarr; Interoperability</p>
---

What happens when there is no standard  
(or, when there are too many standards)

<img src="images/genomic-data-standards/geektech-charging-leads.png" width="600" alt="Geektech charging leads">

---
<img src="images/genomic-data-standards/eu-common-charger.png" width="600" alt="Geektech charging leads">
---

<a href="https://xkcd.com/2116/"><img src="https://imgs.xkcd.com/comics/norm_normal_file_format.png">

---
### Who is the target audience?

- humans?
- computers?

It's easy to build a human-friendly interface to data organized for computers. The opposite is not true.

---
### Aside: My Advice

1. Make and share your data in *computable* form
2. Make *everything* computable
	- literature text processing. we have to get text out of a PDF.

Computability is a prerequisite for interoperability

---
### Existing standards in genomics

- File formats
  - FASTA: DNA sequences
  - FASTQ: Short DNA sequencer reads
  - SAM/BAM: Aligned DNA sequences
  - VCF: genomic variation
  - BED: genomic intervals

---
### File formats are not enough

Say I have two BED files. Are they comparable?

File 1: 
```
chr1  15300  17300
chr1  24900  25600
chr1  72420  74440
```

File 2:
```
1  18110  19300
1  64000  65600
1  72400  74700
```

<div class="fragment">sequence names must match</div>
<div class="fragment">reference sequence must be compatible</div>

---
### Existing standards in genomics

- Reference genomes
  - human reference: hg18, hg19, hg38/GRCh38
  - human T2T reference
  - mouse reference: mm8, mm9, mm10
  - Other reference genomes
---
### Liftover

<img src="images/genomic-data-standards/liftover.jpeg" height="500">

See [Luu et al. 2020](https://doi.org/10.1093/nargab/lqaa054)

---
### Existing standards in genomics
- Gene names
  - HUGO Gene Nomenclature Committee (https://www.genenames.org/)
- Gene and transcript annotation
  - GENCODE (https://www.gencodegenes.org/)

---
### RNA-seq

<a href="https://en.wikipedia.org/wiki/RNA-Seq"><img src="https://upload.wikimedia.org/wikipedia/commons/f/f3/Summary_of_RNA-Seq.svg" height="600"></a>

---
### Transcript models

![](images/genomic-intervals/brca2-gene-model.png)

---

tximeta

---
### GA4GH

> The Global Alliance for Genomics and Health (GA4GH) is a policy-framing and technical standards-setting organization, seeking to enable responsible genomic data sharing within a human rights framework.

- [ga4gh.org](https://www.ga4gh.org/)
- formed in 2013

---
### The increasing importance of web standards

As data and analysis become larger and more distributed, more recent standards increase involvement of web concepts

---
### Application programming interface (API)

- An API defines protocols and formats for input and output of a computing system
- It provides an abstraction layer to a user

---
### REST

- "Representational state transfer"
- REST is a software architectural style.
- It specifies constraints for an API. Most important:

1. Loose coupling between client and server
2. Abstraction of server resources
3. Persistent Uniform Resource Identifiers (URIs)
4. Stateless -- client manages state
5. Representations of resource as universal as possible

---
### JSON

- JavaScript Object Notation is language-agnostic
- The standard interchange format for the web

An object is an unordered set of name/value pairs. An object begins with `{` and ends with `}`. Each name is followed by `:` and the name/value pairs are separated by `,`

```json
{
	"name":"John",
	"age":30,
	"car":null
}

```

---
### Refget

---
### Sequence Collections

---
### GA4GH:DRS

[Data Repository Service (DRS)](https://ga4gh.github.io/data-repository-service-schemas/preview/release/drs-1.0.0/docs/)

Objects in the repository have unique identifiers  

```
/objects/{object_id}
# returns a JSON response including access methods 
```

```
/objects/{object_id}/access/{access_id}
# returns a URL where you can get the object  
``` 

---
### GA4GH:DRS

#### Current approach 

You share your data with a link on your website

<div class="fragment">

#### Future approach

You implement DRS. Then, a standard DRS client could get data, making your data more accessible.

</div>

---
### WES

[Workflow Execution Service](https://ga4gh.github.io/workflow-execution-service-schemas/docs/)

Provides a standard way for users to submit workflow requests to workflow execution systems, and to monitor their execution

- GET `/runs` - to list current runs
- POST `/runs` - to start a new run
- GET `/runs/{run_id}` - information about a run
- GET `/runs/{run_id}/status` - status of a run
- POST `/runs/{run_id}/cancel` - cancels a run

---
### GA4GH:TES

[Task Execution Service](https://ga4gh.github.io/task-execution-schemas/docs/)

- GET `/tasks` - list tasks tracked by the TES server
- POST `/tasks` - create new task
- GET `/tasks/{id}` - get a single task
- POST `/tasks/{id}:cancel` - cancel a task

---

<img src="https://imgs.xkcd.com/comics/standards.png" width="600">

---
### Standards in genomics are important
- Genomes will always be a community resource
- It's a huge, worldwide, rapidly growing endeavor
---
### You should
- Be familiar with existing standards
- Use them whenever possible
- Be aware of work on new standards

<div class="fragment">
Remember, interoperability means impact
</div>