---
layout: reveal_markdown
title: "Genomic data standards"
tags: slides 
date: 2021-12-09
---

<style>


.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:18px;
  overflow:hidden;padding:10px 5px;word-break:normal;}
.tg th{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;
  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}

</style>



# {{ page.title }}
---

## What is a standard?

> A norm or specification applied to a repeated process. It includes a formal definition of terms and a description of formats and expected behavior of a system

<p class="fragment">Standards &rarr; Interoperability</p>
---

What happens when there is no standard  
(or, equivalently, too many standards)

<img src="images/genomic-data-standards/geektech-charging-leads.png" width="600" alt="Geektech charging leads">

---
<img src="images/genomic-data-standards/eu-common-charger.png" width="600" alt="Geektech charging leads">
---

<a href="https://xkcd.com/2116/"><img src="https://imgs.xkcd.com/comics/norm_normal_file_format.png">

---
### Aside: My Advice

Who is the target audience?

- humans?
- computers?

<div class="fragment">It's easy to build a human-friendly interface to data organized for computers. The opposite is not true.</div>

<div class="fragment">
So, always build in this direction: 

computers &rightarrow; humans

</div>

---
### Aside: My Advice

1. Make and share your data in *computable* form
2. Make *everything* computable
	- literature text processing. we have to get text out of a PDF.

Computability is a prerequisite for interoperability




---
### Existing standards in genomics

- File formats
- Reference genomes
- Web standards and APIs
- Ontologies

---
### GA4GH

> The Global Alliance for Genomics and Health (GA4GH) is a policy-framing and technical standards-setting organization, seeking to enable responsible genomic data sharing within a human rights framework.

- [ga4gh.org](https://www.ga4gh.org/)
- formed in 2013

---
### Standards: File formats
  - BED: genomic intervals
  - FASTA: DNA sequences
  - FASTQ: Short DNA sequencer reads
  - SAM/BAM: Aligned DNA sequences
  - VCF: genomic variation

---
### File formats are not enough

Say I have two BED files. Are they comparable?

File 1: 
```
chr1  15300  17300
chr1  24900  25600
chr1  72420  74440
```

File 2:
```
1  18110  19300
1  64000  65600
1  72400  74700
```

<div class="fragment">sequence names must match</div>

<div style="color:black">.</div>
---
### File formats are not enough

Say I have two BED files. Are they comparable?

File 1: 
```
chr1  15300  17300
chr1  24900  25600
chr1  72420  74440
```

File 2:
```
chr1  18110  19300
chr1  64000  65600
chr1  72400  74700
```

sequence names must match. How about now?

<div class="fragment">reference sequence must be compatible</div>
---
<div class="row" style="border: 1px solid white">
	<div class="col" style="border: 1px solid white"></div>
	<div class="col" style="border: 1px solid white">Names Match</div>
	<div class="col" style="border: 1px solid white">Names Mismatch</div>
</div>
<div class="row" style="border: 1px solid white">
	<div class="col" style="border: 1px solid white">Reference Match</div>
	<div class="col" style="border: 1px solid white">Great!</div>
	<div class="col" style="border: 1px solid white">Adjust names?</div>
</div>
<div class="row" style="border: 1px solid white">
	<div class="col" style="border: 1px solid white">Ref. Mismatch</div>
	<div class="col" style="border: 1px solid white">Don't be fooled!</div>
	<div class="col" style="border: 1px solid white">No compatibility</div>
</div>
---
### Standards: Reference genomes

Reference genomes are versioned.

human reference: hg18, hg19, hg38/GRCh38, T2T

mouse reference: mm8, mm9, mm10

They are not compatible!

---
### Converting among genome builds

Liftover ([Luu et al. 2020](https://doi.org/10.1093/nargab/lqaa054))

<img src="images/genomic-data-standards/liftover.jpeg" height="500">

---
### Standards: Reference genomes

There are also many variations of the *same* version.

- 3 GRCh38 providers: NCBI, Ensembl, UCSC
- hard- soft-, or no repeat masks?
- what are the chromosomes named (1 vs chr1 vs chr1, NC_000001.11)
- what secondary scaffolds are included?
- how is the assembly named (hg38, GRCh38, or GCF_000001405.39)
- are any decoy sequences included (like EBV)?

<div class="fragment">
Downstream results are not compatible if based on a different reference!
</div>

---
Andy Yates' "Genome provider analysis"
<table class="tg" style="margin-bottom:60px">
<thead>
  <tr>
    <th class="tg">Provider</th>
    <th class="tg">Chr1 name</th>
    <th class="tg">Chr1 length</th>
    <th class="tg">Chr1 md5</th>
    <th class="tg">Num chroms</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td class="tg">Ensembl primary</td>
    <td class="tg">1</td>
    <td class="tg">248956422</td>
    <td class="tg">2648ae1bacce4ec4b6cf337dcae37816</td>
    <td class="tg">195</td>
  </tr>
  <tr>
    <td class="tg">Ensembl toplevel</td>
    <td class="tg">1</td>
    <td class="tg">248956422</td>
    <td class="tg">2648ae1bacce4ec4b6cf337dcae37816</td>
    <td class="tg">649</td>
  </tr>
  <tr>
    <td class="tg">NCBI</td>
    <td class="tg">NC_000001.11</td>
    <td class="tg">248956422</td>
    <td class="tg">6aef897c3d6ff0c78aff06ac189178dd</td>
    <td class="tg">640</td>
  </tr>
  <tr>
    <td class="tg">UCSC</td>
    <td class="tg">chr1</td>
    <td class="tg">248956422</td>
    <td class="tg">2648ae1bacce4ec4b6cf337dcae37816</td>
    <td class="tg">456</td>
  </tr>
</tbody>
</table>

<div class="small">https://gist.github.com/andrewyatz/692f81baab1bebaf09c481937f2ad6c6</div>


---
### Hashing

> A hash function maps an object to a finite space

- must be deterministic
- should be very fast to compute
- should reduce collisions

<div class="fragment">Used for hash tables, but also...</div>

---
### Hashing

The output, called a *hash* or *digest*, can be seen as a fingerprint or barcode. It is a unique identifier of an input item.

<div class="fragment">As long as there are no collisions...</div>

---
### md5 algorithm

"Message Digest"

---
### 

In 2011, the Internet Engineering Task Force (IETF) [advised in this document](https://tools.ietf.org/search/rfc6151) that

> The published attacks against MD5 show that it is not prudent to use MD5 when collision resistance is required.

---

<h2>How refget works</h2>

<div class="col2">
<img src="images/genomic-data-standards/refget_concept.svg" style="background:white" width="400">
</div>

---
### Sanitize sequence before digesting

1. Remove whitespace
2. Make upper-case

---
### CRAM format


---
Refget is great, but:

- what about *collections* of sequences, like a reference genome?
- what about *names* of sequences?

---

<h2>Sequence collections</h2>

Digest algorithm

<img src="images/genomic-data-standards/digest_algorithm.svg" style="background:white" width="900">

---

<h2>Sequence collections</h2>

Digest lookup

<img src="images/genomic-data-standards/digest_lookup.svg" style="background:white" width="900">

---
### Existing standards in genomics
- Gene names
  - HUGO Gene Nomenclature Committee (https://www.genenames.org/)
- Gene and transcript annotation
  - GENCODE (https://www.gencodegenes.org/)

---
### RNA-seq

<a href="https://en.wikipedia.org/wiki/RNA-Seq"><img src="https://upload.wikimedia.org/wikipedia/commons/f/f3/Summary_of_RNA-Seq.svg" height="600"></a>

---
### Transcript models

![](images/genomic-intervals/brca2-gene-model.png)

---

tximeta

---
### The increasing importance of web standards

As data and analysis become larger and more distributed, more recent standards increase involvement of web concepts

---
### Application programming interface (API)

- An API defines protocols and formats for input and output of a computing system
- It provides an abstraction layer to a user

---
### REST

- "Representational state transfer"
- REST is a software architectural style.
- It specifies constraints for an API. Most important:

1. Loose coupling between client and server
2. Abstraction of server resources
3. Persistent Uniform Resource Identifiers (URIs)
4. Stateless -- client manages state
5. Representations of resource as universal as possible

---
### JSON

- JavaScript Object Notation is language-agnostic
- The standard interchange format for the web

An object is an unordered set of name/value pairs. An object begins with `{` and ends with `}`. Each name is followed by `:` and the name/value pairs are separated by `,`

```json
{
	"name":"John",
	"age":30,
	"car":null
}

```

---
### Refget

---
### Sequence Collections

---
### GA4GH:DRS

[Data Repository Service (DRS)](https://ga4gh.github.io/data-repository-service-schemas/preview/release/drs-1.0.0/docs/)

Objects in the repository have unique identifiers  

```
/objects/{object_id}
# returns a JSON response including access methods 
```

```
/objects/{object_id}/access/{access_id}
# returns a URL where you can get the object  
``` 

---
### GA4GH:DRS

#### Current approach 

You share your data with a link on your website

<div class="fragment">

#### Future approach

You implement DRS. Then, a standard DRS client could get data, making your data more accessible.

</div>

---
### WES

[Workflow Execution Service](https://ga4gh.github.io/workflow-execution-service-schemas/docs/)

Provides a standard way for users to submit workflow requests to workflow execution systems, and to monitor their execution

- GET `/runs` - to list current runs
- POST `/runs` - to start a new run
- GET `/runs/{run_id}` - information about a run
- GET `/runs/{run_id}/status` - status of a run
- POST `/runs/{run_id}/cancel` - cancels a run

---
### GA4GH:TES

[Task Execution Service](https://ga4gh.github.io/task-execution-schemas/docs/)

- GET `/tasks` - list tasks tracked by the TES server
- POST `/tasks` - create new task
- GET `/tasks/{id}` - get a single task
- POST `/tasks/{id}:cancel` - cancel a task

---

<img src="https://imgs.xkcd.com/comics/standards.png" width="600">

---
### Standards in genomics are important
- Genomes will always be a community resource
- It's a huge, worldwide, rapidly growing endeavor
---
### You should
- Be familiar with existing standards
- Use them whenever possible
- Be aware of work on new standards

<div class="fragment">
Remember, interoperability means impact
</div>