---
layout: reveal_markdown
title: "Genomic Intervals"
tags: slides 
date: 2021-12-09
---

<style>
	.reveal {font-size: 2.2em;}
</style>

## {{ page.title }}
---
## Terminology

- **interval**: a location on a sequence defined by a *start* and *end*
- **region**: often used as a synonym of *interval*
- **chromosome**: In the context of genomic intervals, *chromosome* is often used to indicate the *name of the sequence* on which an interval is defined, but this is really a fudging
- **interval set** or **region set**: a set of intervals/regions

---
<img src="images/genomic-intervals/intervals.svg">

---
## What can be represented as an interval?

- ChIP-seq or ATAC-seq peaks

---
## Peaks

![](images/genomic-intervals/peaks.svg)

Genomic intervals are often colloquially referred to as 'peaks'.

---
## What can be represented as an interval?

- ChIP-seq or ATAC-seq peaks
- Single-Nucleotide Polymorphisms (SNPs) 

---
## SNPs

![](images/genomic-intervals/snps.svg)

SNPs are interval of width 1

---
## What can be represented as an interval?

- ChIP-seq or ATAC-seq peaks
- Single-Nucleotide Polymorphisms (SNPs)
- Genes and gene components (TSS, exons, introns, etc)

---
## Genes and gene components

![](images/genomic-intervals/brca2-gene-model.png)

---
## What can be represented as an interval?

- ChIP-seq or ATAC-seq peaks
- Single-Nucleotide Polymorphisms (SNPs)
- Genes and gene components (TSS, exons, introns, etc)
- Non-coding DNA annotation (promoters, enhancers)

---
## Non-coding DNA annotation

![](images/genomic-intervals/regulatory-build.png)

---
## What can be represented as an interval?

- ChIP-seq or ATAC-seq peaks
- Single-Nucleotide Polymorphisms (SNPs)
- Genes and gene components (TSS, exons, introns, etc)
- Non-coding DNA annotation (promoters, enhancers)
- Any sequence or subsequence (*e.g.* a chromosome)
- Chromosomes

---
![](images/genomic-intervals/chrom-sizes.png)

---
## Terminology

- **coordinate system**: a set of sequence names and lengths


---
## What can be represented as an interval?

- ChIP-seq or ATAC-seq peaks
- Single-Nucleotide Polymorphisms (SNPs)
- Genes and gene components (TSS, exons, introns, etc)
- Non-coding DNA annotation (promoters, enhancers)
- Any sequence or subsequence (*e.g.* a chromosome)
- Chromosomes
- Protein domains

---
## Protein domains

![](images/genomic-intervals/ets-domains.jpeg)

---
## What can be represented as an interval?

- ChIP-seq or ATAC-seq peaks
- Single-Nucleotide Polymorphisms (SNPs)
- Genes and gene components (TSS, exons, introns, etc)
- Non-coding DNA annotation (promoters, enhancers)
- Any sequence or subsequence (*e.g.* a chromosome)
- Chromosomes
- Protein domains
- Anything else?

---
# Key point

<span class="fragment">
Because of the linear nature of DNA and RNA, many biological entities can be conceptualized as genomic intervals.
</span>
<span class="fragment">Genomic intervals are often a simplified abstraction of genomic sequence.</span>

<span class="fragment">
$ \implies $
Interval operations are fundamental in genomics
</span>
---
## File formats for genomic intervals
- [BED](http://genome.ucsc.edu/FAQ/FAQformat.html#format1)
- [narrowPeak](http://genome.ucsc.edu/FAQ/FAQformat.html#format12)
- [broadPeak](http://genome.ucsc.edu/FAQ/FAQformat.html#format13)
- Genome Feature Format ([GTF](http://uswest.ensembl.org/info/website/upload/gff.html), [GFF2](http://gmod.org/wiki/GFF2), [GFF3](https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md))
  - Used for storing transcripts, exons, *etc.*
- Many custom variations!

---
### BED (Browser Extensible Data)

- originally designed by UCSC for the genome browser
- but, also often used for analysis
- a TSV file with one row per genomic region
- 3 required columns: `chrom`, `chromStart`, and `chromEnd`

Example of BED3 file:
```
chr1      15      250
chr2   38252    49502
```

---
## BED

- 9 additional optional fields:
	- name: name of region
	- score: range 0-1000
	- strand: `+`, `-`, or `.`
	- thickStart
	- thickEnd
	- itemRgb: display color (e.g. `55,0,0`)
	- blockCount
	- blockSizes
	- blockStarts

---
## BED

- BED3 is a BED with the first 3 columns
- BED6 is `chrom`, `chromStart`, and `chromEnd`, `name`, `score`, `strand`

---
##  Related file formats
- [Wiggle](http://genome.ucsc.edu/goldenPath/help/wiggle.html) - continuous-valued data with constant size
	- Fixed-step - regular spacing between
	- Variable-step - variable spacing between
- BedGraph - continuous-valued data with variable size

---
## Wiggle (fixed step)

Continuous-valued data with constant size and regular spacing

Format:
```
fixedStep  chrom=chrN start=position  step=stepInterval [span=windowSize]
dataValue1
dataValue2
...
```

Example
```
fixedStep chrom=chr3 start=400601 step=100
11
22
33
```

---
## Wiggle (variable step)

Continuous-valued data with constant size and variable spacing


Format:
```
variableStep  chrom=chrN [span=windowSize]
  chromStartA  dataValueA
  chromStartB  dataValueB
  ... etc ...  ... etc ...
```

Example:
```
variableStep chrom=chr2
300701 12.5
300702 12.5
300703 12.5
300704 12.5
300705 12.5 
```

---
## BedGraph

Continuous-valued data with variable size

```
track type=bedGraph
chr19 49302000 49302300 -1.0
chr19 49302300 49302600 -0.75
chr19 49302600 49302900 -0.50
chr19 49302900 49303200 -0.25
chr19 49303200 49303500 0.0
chr19 49303500 49303800 0.25
```

---
## Terminology

- **intersection**
- **union**
- **overlap**
- **flat**: a region set that is non-overlapping

---
### Biological questions, interval computations

- Is SNP $s$ located in a promoter element?
- Is SNP $s$ located in a gene-dense region?
- What is the nearest gene downstream of SNP $s$?
- What is the average distance to genes for SNP set $S$?

---
### Biological questions, interval computations

- Which TF binding sizes overlap region $r$?
- Are binding sites of TF $y$ enriched in enhancers?
- How many sites of TF $y$ are within 10 kb of gene $z$?
- Has any previous experiment previously identified region $r$?
- What gene has the greatest concentration of TF $y$ binding sites within 1kb?

---
### Biological questions, interval computations
- What is the average signal across interval set $I$?
- What is the *shape* of signal across interval set $I$?
- What is the frequency of annotations for interval set $I$?
- How many SNPs in patient $p$ overlap interval set $I$?

---
### Biological questions, interval computations
- How many aligned reads mapped to gene $g$?
- How many aligned reads overlap splicing site $s$?

---
## What constraints make two intervals comparable?

- An interval comparison assumes a homology statement.


---
# Intersection

Does $A$ intersect $B$?

$  A \overset{?}\cap B$

![](images/genomic-intervals/intervals-intersect.svg)


---
![](images/genomic-intervals/intervals-intersect.svg)

$A_{start} = 5$ &nbsp; $A_{end} = 25$ &nbsp; $B_{start} = 23$ &nbsp; $B_{end} = 30$

$cond 1: A_{end} \overset{?}> B_{start}$

<span class="fragment">
$cond 2: A_{start} \overset{?}< B_{end}$
</span>

---
Assume:
1. Intervals are well-formed ($X_{start} < X_{end}$)
2. Intervals are ordered by start ($A_{start} < B_{start}$)

$A_{start} < B_{start}  $ (order)  
$B_{start} < B_{end}  $  (well-formed)  
$\implies A_{start} < B_{end}$ (trans. property)

Then:

if $ A_{end} > B_{start}$

$ \implies A \cap B$


---
## Union

1. Do they intersect? Take min/max

---
###

If we make some assumpions, like pre-computing order, we can do things more efficiently.

---
### Interval set overlap counting

Given:

A database intervals: $i_1,i_2,...\in\textbf{I}$,   
A query interval: $q$,  

Count the number of intervals in $\textbf{I}$ overlapping $q$.

<pre><code data-line-numbers="1-9">def count_overlaps(query, intervals_set):
	n_overlaps = 0
	for i in interval_set:
		if overlap(query, i):
			n_overlaps = n_overlaps +1
	return n_overlaps

def overlap(query, i):
	return query.start < i.end && query.end > i.start
</code></pre>

Application: What is the RNA-seq expression level of gene $g$?

---

<pre><code data-line-numbers="3">def count_overlaps(query, intervals_set):
	n_overlaps = 0
	for i in interval_set:
		if overlap(query, i):
			n_overlaps = n_overlaps +1
	return n_overlaps

def overlap(query, i):
	return query.start < i.end && query.end > i.start
</code></pre>


$O(N)$  
$N$ = Number of intervals in database

We may have billions of intervals. 

Can we do better?

---
## Binary search

```Python
def binary_search(array, query, low, high):
	pivot = int((low + high)/2)
	if (low >= high):
		return False
	if (array[pivot] = query):
		return pivot
	if (array[pivot] > query):
		return binary_search(array, query, low, pivot-1)
	else:
		return binary_search(array, query, pivot+1, high)
```
> Binary search is one of the most powerful ideas in algorithm design... After only one comparison you can discard one half of all the names in the book. Twenty comparisons suffice to find any name in the million-name Manhattan phone book. -*Steven Skiena, The Algorithm Design Manual*
---
## Binary search

- Sequential search: $O(N)$; Binary search: $O(logN)$
- Requires sorted input (sequential search does not)

Say I have a series of regions in a BED file. Can I do a binary search on the file?

<div class="fragment">
Binary search also requires random access.

> Random access means the ability to access any element directly. This is contrast to *sequential access*, where I can only access elements in sequence.
</div>

---
### Binary search trees

<div class="mermaid">
graph TB;
    A((8))-->B((3))
    A-->C((10))
    B-->D((2))
    B-->E((6))
    C-->F((9))
    C-->G((12))
    E-->H((5))
    E-->I((7))
</div>


- Random access and array *insertion* are enemies.
- Insertion in a sorted array is $O(N)$.
- Binary search **trees** have $O(logN)$ search *and*  insertion.
- They do not require random access for binary search.

---
An $O(logN)$ option. Organize database as two lists:

![](images/genomic-intervals/subject-query-annotated.svg)

<div class="fragment">

1. Binary search $q_{end}$ vs $\textbf{I}_{start}$  
2. Step backward until no more overlaps  

![](images/genomic-intervals/binary-search.svg)
</div>



---

Hidden assumption: No containment

![](images/genomic-intervals/subject-query-containment.svg)

![](images/genomic-intervals/binary-search-fail.svg)

<div class="fragment">
How can we handle containment?
</div>

---
## Advanced data structures

- B trees
- R trees
- B+ trees
- Interval trees
- Nested containment lists

---
### B trees

- [B Tree at Wikipedia](https://en.wikipedia.org/wiki/B-tree)
- Generalization of a binary search tree:
	- Nodes have $m$ children
	- Node values are *boundaries* for subtrees
	- Boundaries provide overlap constraints

![](images/genomic-intervals/B-tree.svg)

---
### R trees
- Generalization of a B-tree beyond integer keys (used by BEDtools)
- A flavor of [k-d trees](https://en.wikipedia.org/wiki/K-d_tree)
- Annotates tree nodes with a *minimum bounding rectangle* of elements. A query that does not intersect the bounding rectangle will not intersect any child element.
- Introduced by [Guttman 1984](https://doi.org/10.1145/971697.602266)

---

<img src="images/genomic-intervals/R-tree.png" height=650>

---
### Nested containment list

- Used by GenomicRanges
![](images/genomic-intervals/nclist.png)

---
### B+ trees

- A B tree in which nodes hold keys pointing to a linked list of values.
- Used by GIGGLE
- See [B+ Tree at Wikipedia](https://en.wikipedia.org/wiki/B%2B_tree)

![](images/genomic-intervals/Bplustree.png)

---
### Region set enrichment analysis

Given a query region set (say, from a ChIP-seq experiment), what other, previously published region sets are most similar?


---
### Interval similarity metrics
- Overlap count
- Jaccard index $\frac{A \cap B}{A \cup B}$

---
<img src="images/genomic-intervals/07-lola1.svg" />
---
<img src="images/genomic-intervals/08-lola2.svg" />
---
<img src="images/genomic-intervals/09-test.svg" />
---
<img src="images/genomic-intervals/10-results.svg" />



---
# Conclusion

- Genomic intervals are an extremely useful abstraction.
- Genomic interval data is growing astronomically
- Diverse biological questions can be asked using intervals
- Data structures and algorithms enables large-scale analysis
- New methods for genomic intervals are exciting
